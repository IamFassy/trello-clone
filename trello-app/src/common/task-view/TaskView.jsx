import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { removeItemFromTask, updateItemInTask } from '../../reducers/cardData';
import IconButton from '../icon-button/IconButton';
import './task-view.scss';

const TaskView = ({ task, index, id }) => {
	const dispatch = useDispatch();
	const [inputView, setInputView] = useState(false);
	const removeTask = () => {
		dispatch(removeItemFromTask({ id, taskIndex: index }));
	};

	const showInput = () => {
		setInputView((view) => !view);
	};

	const updateTask = (e) => {
		dispatch(
			updateItemInTask({ id, taskIndex: index, task: e.target.value })
		);
	};

	const dragStart = (e) => {
		const data = JSON.stringify({
			type: 'task',
			taskName: task,
			id,
		});
		e.dataTransfer.setData('text/plain', data);
	};

	const handleDragEnd = (e) => {
		e.dataTransfer.clearData();
	};

	return (
		<div
			className='task-view'
			draggable
			onDragStart={dragStart}
			onDragEnd={handleDragEnd}
		>
			<div className='task-view__left-view'>
				{!inputView ? (
					<p className='task-view__left-view-title'>{task}</p>
				) : (
					<input
						className='task-view__left-view-input'
						value={task}
						onChange={updateTask}
						data-testid='task-input'
						onBlur={showInput}
					/>
				)}
			</div>

			<IconButton onClick={showInput} testId='toggle-input'>
				<FontAwesomeIcon size={'xs'} icon='fa-solid fa-pen' />
			</IconButton>
			<IconButton onClick={removeTask} testId='remove-task'>
				<FontAwesomeIcon
					color={'red'}
					size={'xs'}
					icon='fa-solid fa-xmark'
				/>
			</IconButton>
		</div>
	);
};

export default TaskView;
