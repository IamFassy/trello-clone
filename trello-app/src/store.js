import { configureStore } from '@reduxjs/toolkit';
import cardDataReducer from './reducers/cardData';

export const store = configureStore({
	reducer: {
		cardData: cardDataReducer,
	},
});
