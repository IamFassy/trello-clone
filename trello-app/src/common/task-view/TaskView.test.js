import React from 'react';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import TaskView from './TaskView';
import { Provider } from 'react-redux';
import { store } from '../../store';
import reducer, {
	removeItemFromTask,
	updateItemInTask,
} from '../../reducers/cardData';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faPen, faXmark } from '@fortawesome/free-solid-svg-icons';

library.add(faPen, faXmark);

afterEach(cleanup);

test('should take a snapshot', () => {
	const { asFragment } = render(
		<Provider store={store}>
			<TaskView task='Test' index={0} id={1} />
		</Provider>
	);
	expect(
		asFragment(
			<Provider store={store}>
				<TaskView task='Test' index={0} id={1} />
			</Provider>
		)
	).toMatchSnapshot();
});

test('remove item from task', () => {
	const previousState = {
		cards: [{ id: 1, tasks: ['First Item'], title: 'Change Title' }],
	};
	const arr = {
		cards: [{ id: 1, tasks: [], title: 'Change Title' }],
	};

	render(
		<Provider store={store}>
			<TaskView id={1} task='First Item' index={0} />
		</Provider>
	);
	fireEvent.click(screen.getByTestId('remove-task'));
	expect(
		reducer(previousState, removeItemFromTask({ id: 1, index: 0 }))
	).toEqual(arr);
});

test('show input view', () => {
	render(
		<Provider store={store}>
			<TaskView id={1} task='First Item' index={0} />
		</Provider>
	);
	fireEvent.click(screen.getByTestId('toggle-input'));
	expect(screen.getAllByTestId('task-input')).toBeTruthy();
});

const setup = () => {
	const utils = render(
		<Provider store={store}>
			<TaskView id={1} task='First Item' index={0} />
		</Provider>
	);
	const input = utils.getByTestId('task-input');
	return {
		input,
		...utils,
	};
};

test('update item in task', () => {
	const previousState = {
		cards: [{ id: 1, tasks: ['First Item'], title: 'Change Title' }],
	};
	const arr = {
		cards: [{ id: 1, tasks: ['Updated Item'], title: 'Change Title' }],
	};

	render(
		<Provider store={store}>
			<TaskView id={1} task='First Item' index={0} />
		</Provider>
	);
	fireEvent.click(screen.getByTestId('toggle-input'));
	const { input } = setup();
	fireEvent.change(input, { target: { value: 'Updated Item' } });
	expect(
		reducer(
			previousState,
			updateItemInTask({ id: 1, index: 0, task: 'Updated Item' })
		)
	).toEqual(arr);
});
