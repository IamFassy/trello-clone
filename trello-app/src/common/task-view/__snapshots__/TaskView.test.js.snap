// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="task-view"
    draggable="true"
  >
    <div
      class="task-view__left-view"
    >
      <p
        class="task-view__left-view-title"
      >
        Test
      </p>
    </div>
    <button
      class="icon-button "
      data-testid="toggle-input"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-pen fa-xs "
        data-icon="pen"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M362.7 19.32C387.7-5.678 428.3-5.678 453.3 19.32L492.7 58.75C517.7 83.74 517.7 124.3 492.7 149.3L444.3 197.7L314.3 67.72L362.7 19.32zM421.7 220.3L188.5 453.4C178.1 463.8 165.2 471.5 151.1 475.6L30.77 511C22.35 513.5 13.24 511.2 7.03 504.1C.8198 498.8-1.502 489.7 .976 481.2L36.37 360.9C40.53 346.8 48.16 333.9 58.57 323.5L291.7 90.34L421.7 220.3z"
          fill="currentColor"
        />
      </svg>
    </button>
    <button
      class="icon-button "
      data-testid="remove-task"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-xmark fa-xs "
        color="red"
        data-icon="xmark"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M310.6 361.4c12.5 12.5 12.5 32.75 0 45.25C304.4 412.9 296.2 416 288 416s-16.38-3.125-22.62-9.375L160 301.3L54.63 406.6C48.38 412.9 40.19 416 32 416S15.63 412.9 9.375 406.6c-12.5-12.5-12.5-32.75 0-45.25l105.4-105.4L9.375 150.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L160 210.8l105.4-105.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-105.4 105.4L310.6 361.4z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
